polyply/src/annotate_dna.py

```diff
@@ -0,0 +1,194 @@
+                "DA3": "DT3", "DT3": "DA3", "DG3": "DC3", "DC3": "DG3"
+                }
+
+class AnnotateDNA(Processor):
+    """
+    Given a topology and list indexes that indicate the molecules that
+    contain DNA, the processor generates the appropriate meta_molecule
+    structure. The DNA strands are identified using a greedy search method
+    on the meta_molecule network and later combined with their complementary
+    strand. The generated meta_molecule is later used for both building the
+    DNA structure and backmapping to the desired forcefield.
+    """
+
+    def __init__(self, topology, includes_DNA):
+        """
+        Initalize the processor with a :class:`topology`
+        and a list of DNA molecules in this topology.
+
+        Parameters
+        ------------
+        topology: :class: `polyply.src.topology.Topology`
+        includes_DNA: list[int]
+            node keys list of the meta_molecules tagged as indluding DNA
+        """
+        self.topology = topology
+        self.includes_DNA = [float(ndx) for ndx in includes_DNA]
+
+    def _find_dna_strands(self, meta_molecule):
+        """
+        Extract the subgraphs of meta_molecule that are DNA strands
+
+        Parameters
+        ------------
+        meta_molecule: :class:`polyply.src.MetaMolecule`
+
+        Returns
+        ---------
+        strands: list[:class: `networkx.Graph`]
+            list of DNA strands represented as subgraphs of meta_molecule
+        """
+
+        def _is_nucleobase(node):
+            return (meta_molecule.nodes[node]['resname'] in base_library)
+
+        seperated_meta_mol = [meta_molecule.subgraph(graph) for graph
+                              in nx.connected_components(meta_molecule)]
+
+        # Find all strands by performing greedy search on
+        # all connected_components of meta_molecule
+        strands = []
+        for meta_mol in seperated_meta_mol:
+            queue = sample(meta_mol.nodes, 1)
+            visited = []
+            strand = []
+            while queue:
+                # pop first node out of queue
+                current_node = queue.pop()
+                visited.append(current_node)
+
+                if _is_nucleobase(current_node):
+                    strand.append(current_node)
+                elif strand:
+                    strands.append(meta_mol.subgraph(strand))
+                    strand = []
+
+                # Determine neighbors nodes
+                neighbors = [node for node in meta_molecule.neighbors(current_node) if node not in queue + visited]
+                # update queue
+                queue = neighbors + queue
+                # Sort queue according to heuristic
+                queue.sort(key=_is_nucleobase, reverse=True)
+
+            if strand not in strands:
+                strands.append(meta_molecule.subgraph(strand))
+        return strands
+
+    def _determine_sequence(self, strand):
```

Comment:
So strand is a subgraph of the original molecule? Just do this to get the sequence:
```
resnames = nx.get_node_attributes(subgraph, "resname")
sequence = resnames.values()
```
If you want you can also use the numerical dict keys to order the resnames such that the sequence is ordered. Since your function does not preserve order I assume that is not needed

Reply:
@fgrunewald I thought that doing it this way could lead to problems, due to the way that I recognise DNA strands in a meta_molecule. See diagram below, the numbers indicate the order in which the depth first search visits the residues.
In this case the order in which the strand is read, is not equal to their relative position in the strand. This is why I first look for a terminal node in my strand and then read in the sequence. Am I over-complicating/slowing down the code or could this help robustness in the future? Not entirely sure myself.

![image](https://user-images.githubusercontent.com/49684318/151360784-8e20f656-7137-4b69-806c-d9eabb9ecafe.png)